include/deal.II/lac/linear_operator.h

```diff
@@ -1215,6 +1011,1082 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
+  };
+
+  return_op.Tvmult_add = [ops](Domain &v, const Range &u)
+  {
+    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
+    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+    for (unsigned int i = 0; i < m; ++i)
+      ops[i].Tvmult_add(v.block(i), u.block(i));
+  };
+
+  return return_op;
+}
+
+
+/**
+ * @relates LinearOperator
+ *
+ * A variant of above function that only takes a single LinearOperator
+ * argument @p op and creates a blockdiagonal linear operator with @p m
+ * copies of it.
+ *
+ * @ingroup LAOperators
+ */
+template <unsigned int m,
+          typename Range = BlockVector<double>,
+          typename Domain = Range>
+LinearOperator<Range, Domain>
+block_diagonal_operator(const LinearOperator<typename Range::BlockType, typename Domain::BlockType> &op)
+{
+
+  static_assert(m > 0,
+                "a blockdiagonal LinearOperator must consist of at least "
+                "one block");
+
+  LinearOperator<Range, Domain> return_op;
+
+  return_op.reinit_range_vector = [op](Range &v, bool fast)
+  {
+    // Reinitialize the block vector to m blocks:
+    v.reinit(m);
+
+    // And reinitialize every individual block with reinit_range_vectors:
+    for (unsigned int i = 0; i < m; ++i)
+      op.reinit_range_vector(v.block(i), fast);
+
+    v.collect_sizes();
+  };
+
+  return_op.reinit_domain_vector = [op](Domain &v, bool fast)
+  {
+    // Reinitialize the block vector to m blocks:
+    v.reinit(m);
+
+    // And reinitialize every individual block with reinit_domain_vectors:
+    for (unsigned int i = 0; i < m; ++i)
+      op.reinit_domain_vector(v.block(i), fast);
+
+    v.collect_sizes();
+  };
+
+  return_op.vmult = [op](Range &v, const Domain &u)
+  {
+    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
+    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+    for (unsigned int i = 0; i < m; ++i)
+      op.vmult(v.block(i), u.block(i));
+  };
+
+  return_op.vmult_add = [op](Range &v, const Domain &u)
+  {
+    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
+    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+    for (unsigned int i = 0; i < m; ++i)
+      op.vmult_add(v.block(i), u.block(i));
+  };
+
+  return_op.Tvmult = [op](Domain &v, const Range &u)
+  {
+    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
+    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+    for (unsigned int i = 0; i < m; ++i)
+      op.Tvmult(v.block(i), u.block(i));
+  };
+
+  return_op.Tvmult_add = [op](Domain &v, const Range &u)
+  {
+    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
+    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+    for (unsigned int i = 0; i < m; ++i)
+      op.Tvmult_add(v.block(i), u.block(i));
+  };
+
+  return return_op;
+}
+
+//@}
+
+
+
+/**
+ * A class to store a computation.
```

Comment:
@bangerth If the header file grows beyond the 2200 lines (in total), I'll split it into two files. Currently, I like to keep it as it is because there is not much use of a seperate ```package_operation.h``` header file on the user side.